### PR TITLE
Add homepage posts, donation, and subscribe patterns

### PR DESCRIPTION
**Pull request by thomasguillot**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds:
* 8 homepage posts (HP) block patterns (currently called Style 1 - 8 -- probably needs to change) -- they are all a mix of group, columns, and homepage posts
* 3 donation (D) patterns
* 2 subscribe (S) patterns

__Screencast:__ https://cloudup.com/cmhT1DvoAbg

_Note: HP are only available for pages whereas D and S are available for pages, posts, and pop-ups._

_Note 2: The patterns will require a fix for the pop-ups -- alignfull overflows the pop-up container_

### How to test the changes in this Pull Request:

1. Switch to this branch
2. Try to add different homepage posts patterns to a page.
3. Try to add different subscribe and donation patterns to a posts and pop-up.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by thomasguillot** [comment]
> When there's text in the content, it won't be localized. We'll need a way to tokenize the text so that the pieces of text can be added in the newspack_blocks_patterns filter and they will replace the text in the markup.

Yep. How do we localize the json file?

> Several of the patterns have hardcoded references to a media asset. This won't work unless the site has Newspack starter content. We'll also need a solution where we can tokenize the first asset in the media library, or the largest asset available.

The image used is actually located on newspack.blog: https://newspack.blog/wp-content/uploads/2020/03/automated_upload-62.jpg

> Question: if there are no assets at all, should we not show the patterns that require assets?

See previous answer.

> Question: should we remove __filte and title parameters from the json files since these aren't used anywhere?

I didn't remove them because that's the file automatically generated by the reusable blocks (which is what I'm using to generate the json files). And I kept them in case they are useful once with have Core's pattern library.

> Question: Should the Homepage Posts blocks be set to Full Width alignment?

They were originally full width. Laurel convinced me not to have them full width unless with have a background colour.

> For Donate blocks, we should suppress them if the Donation feature isn't configured.

Yes, we should make sure we don't show patterns with unavailable blocks. Same for the Mailchimp block... Not sure where to add the check though.

**Comment by jeffersonrabb** [comment]
> Yep. How do we localize the json file?

I considered a few options, here's my recommendation. Leave the strings in the pattern markup exactly as is, and use `str_replace` to replace with localized equivalents. In this example I'm using arrays for the search and replace even though there's only one string — this leaves the pattern open to multiple strings, if needed.

```
$decoded = json_decode( file_get_contents( NEWSPACK_BLOCKS__PLUGIN_DIR . 'src/patterns/markup/donation/style-1.json'), true ); //phpcs:ignore
$patterns[] = [
	'category' => __( 'Donation', 'newspack-blocks' ),
	'content'  => str_replace(
		[
			'Support our publication',
		],
		[
			__( 'Support our publication', 'newspack-blocks' ),
		],
		$decoded['content']
	),
	'image'    => plugins_url( 'src/patterns/images/donation/style-1.png', __FILE__ ),
	'title'    => __( 'Style 1', 'newspack-blocks' ),
];
```

**Comment by jeffersonrabb** [comment]
> I didn't remove them because that's the file automatically generated by the reusable blocks (which is what I'm using to generate the json files). And I kept them in case they are useful once with have Core's pattern library.

Good thinking, agreed.

**Comment by jeffersonrabb** [comment]
> Yes, we should make sure we don't show patterns with unavailable blocks. Same for the Mailchimp block... Not sure where to add the check though.

Example of checking if the Donate block is configured:

```
add_filter(
	'newspack_blocks_patterns',
	function( $patterns, $post_type ) {
		if ( in_array( $post_type, [ 'page', 'post', 'newspack_popups_cpt' ], true ) ) {
			$newspack_donations_configured = false;
			if ( class_exists( 'Newspack\Donations' ) ) {
				$settings = Newspack\Donations::get_donation_settings();
				if ( ! is_wp_error( $settings ) && $settings['created'] ) {
					$newspack_donations_configured = true;
				}
			}
			if ( $newspack_donations_configured ) {
				$from_json = json_decode( file_get_contents( NEWSPACK_BLOCKS__PLUGIN_DIR . 'src/patterns/markup/donation/style-1.json'), true ); //phpcs:ignore
				$patterns[] = [
					'category' => __( 'Donation', 'newspack-blocks' ),
					'content'  => $from_json['content'],
					'image'    => plugins_url( 'src/patterns/images/donation/style-1.png', __FILE__ ),
					'title'    => __( 'Style 1', 'newspack-blocks' ),
				];
			}
		}
		return $patterns;
	},
	10,
	2
);
``` 

Example of checking if Mailchimp is configured:

```
add_filter(
	'newspack_blocks_patterns',
	function( $patterns, $post_type ) {
		if ( in_array( $post_type, [ 'page', 'post', 'newspack_popups_cpt' ], true ) ) {
			if ( ! function_exists( 'jetpack_mailchimp_verify_connection' ) || ! jetpack_mailchimp_verify_connection() ) {
				return $patterns;
			}
			$from_json = json_decode( file_get_contents( NEWSPACK_BLOCKS__PLUGIN_DIR . 'src/patterns/markup/subscribe/style-1.json'), true ); //phpcs:ignore
			$patterns[] = [
				'category' => __( 'Subscribe', 'newspack-blocks' ),
				'content'  => $from_json['content'],
				'image'    => plugins_url( 'src/patterns/images/subscribe/style-1.png', __FILE__ ),
				'title'    => __( 'Style 1', 'newspack-blocks' ),
			];
		}
		return $patterns;
	},
	10,
	2
);
```

File: src/patterns/markup/homepage-posts/style-5.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
...
...lumn -->\n\n<!-- wp:column -->\n<div class=\"wp-block-column\"><!-- wp:newspack-blocks/homepage-articles {\"className\":\"is-style-borders\",\"showExcerpt\":false,\"showAvatar\":false,\"postsToShow\":1,\"categories\":[],\"typeScale\":3,\"imageScale\":1} /-->\n\n<!-- wp:newspack-blocks/homepage-articles {\"className\":\"is-style-borders\",\"showExcerpt\":false,\"showImage\":false,\"showAuthor\":false,\"showAvatar\":false,\"mediaPosition\":\"left\",\"categories\":[],\"typeScale\":1,\"imageScale\":1} /--></div>\n<!-- /wp:column -->\n\n<!-- wp:column -->\n<div class=\"wp-block-column\"><!-- wp:newspack-blocks/homepage-articles {\"className\":\"is-style-borders\",\"showExcerpt\":false,\"showAvatar\":false,\"postsToShow\":1,\"categories\":[],\"typeScale\":3,\"imageScale\":1,\"mobileStack\":true} /-->\n\n<!-- wp:newspack-blocks/homepage-articles {\"className\":\"is-style-borders\",\"showExcerpt\":false,\"showImage\":false,\"showAuthor\":false,\"showAvatar\":false,\"mediaPosition\":\"left\",\"categories\":[],\"typeScale\":1,\"imageScale\":1} /-->\n\n<!-- wp:paragraph -->\n<p></p>\n<!-- /wp:paragraph --></div>\n<!-- /wp:column --></div>\n<!-- /wp:columns --></div></div>\n<!-- /wp:group -->"
```
Comment (laurelfulford) [inline src/patterns/markup/homepage-posts/style-5.json]: There is a rogue empty paragraph block at the veeeeery end of this line:

`<!-- wp:paragraph -->\n<p></p>\n<!-- /wp:paragraph -->`

File: src/patterns/markup/homepage-posts/style-2.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
...\"postsToShow\":1,\"typeScale\":6} /-->\n\n<!-- wp:paragraph -->\n<p></p>\n<!-- /wp:paragraph --></div>\n<!-- /wp:column -->\n\n<!-- wp:column {\"width\":25} -->\n<div class=\"wp-block-column\" style=\"flex-basis:25%\"><!-- wp:newspack-blocks/homepage-articles {\"showExcerpt\":false,\"minHeight\":25,\"showAvatar\":false,\"postsToShow\":1,\"mediaPosition\":\"behind\",\"typeScale\":2} /-->\n\n<!-- wp:newspack-blocks/homepage-articles {\"className\":\"is-style-borders\",\"showExcerpt\":false,\"showImage\":false,\"showAvatar\":false,\"postsToShow\":5,\"typeScale\":2,\"imageScale\":1} /--></div>\n<!-- /wp:column -->\n\n<!-- wp:column {\"width\":25} -->\n<div class=\"wp-block-column\" style=\"flex-basis:25%\"><!-- wp:newspack-blocks/homepage-articles {\"showExcerpt\":false,\"minHeight\":25,\"showAvatar\":false,\"postsToShow\":1,\"mediaPosition\":\"behind\",\"typeScale\":2} /-->\n\n<!-- wp:newspack-blocks/homepage-articles {\"className\":\"is-style-borders\",\"showExcerpt\":false,\"showImage\":false,\"showAvatar\":false,\"postsToShow\":5,\"mediaPosition\":\"left\",\"typeScale\":2,\"imageScale\":1} /--></div>\n<!-- /wp:column --></div>\n<!-- /wp:columns --></div></div>\n<!-- /wp:group -->"
```
Comment (laurelfulford) [inline src/patterns/markup/homepage-posts/style-2.json]: Minor nitpick, but there's an empty paragraph block at the end of the first column (the focus currently ends up in that column anyway, which would also create an empty paragraph block, so removing it won't seem to do anything for now):

`<!-- wp:paragraph -->\n<p></p>\n<!-- /wp:paragraph -->`

File: src/patterns/markup/homepage-posts/style-5.json (hunk 1 above)
Comment (thomasguillot) [reply src/patterns/markup/homepage-posts/style-5.json]: Well spotted! You passed the test 😜

Fixed in https://github.com/Automattic/newspack-blocks/pull/387/commits/14a9e3da1a12a47856260c0d7ceac1f5809c6508

File: src/patterns/markup/homepage-posts/style-2.json (hunk 2 above)
Comment (thomasguillot) [reply src/patterns/markup/homepage-posts/style-2.json]: https://github.com/Automattic/newspack-blocks/pull/387/commits/14a9e3da1a12a47856260c0d7ceac1f5809c6508

**Comment by thomasguillot** [comment]
> When you add a pattern, you can end up with one or two empty paragraph blocks. I ended up with one above my new block when I had my cursor on an empty line -- rather than replacing the empty line, the pattern ended up below it:

> When adding patterns with columns, the cursor ended up inside of the first column, rather than focus on the whole block

> Is it possible to make sure patterns are at least always added below?

I have noticed this too.

cc @jeffersonrabb for ideas/thoughts

**Comment by thomasguillot** [comment]
> Leave the strings in the pattern markup exactly as is, and use `str_replace` to replace with localized equivalents

Done in https://github.com/Automattic/newspack-blocks/pull/387/commits/d3daf0c566dec08b362bddf4deeaed43af5dd35f

**Comment by thomasguillot** [comment]
> Example of checking if the Donate block is configured:

> Example of checking if Mailchimp is configured:

✅  https://github.com/Automattic/newspack-blocks/pull/387/commits/371d5837a0f15da06804b5289bddf94705a6cf6b

File: src/patterns/markup/homepage-posts/style-5.json (hunk 1 above)
Comment (laurelfulford) [reply src/patterns/markup/homepage-posts/style-5.json]: Haha yay! 😆 

(Both of these fixes look good, thanks Thomas!)

**Comment by jeffersonrabb** [comment]
> When you add a pattern, you can end up with one or two empty paragraph blocks. I ended up with one above my new block when I had my cursor on an empty line -- rather than replacing the empty line, the pattern ended up below it:

I made some changes in https://github.com/Automattic/newspack-blocks/pull/387/commits/598591168ada47998bada7ad97997e8ecbd5262c to address this. Originally, the pattern was inserted immediately after the selected block. With this change, if the selected block is an empty paragraph, the pattern will replace it. I wasn't able to replicate the column block selection issue, but it's possible this will fix it too.

@thomasguillot other than this I think we're good to go. If you can verify this fix, I'll approve and we can merge.

**Comment by thomasguillot** [comment]
> I made some changes in [5985911](https://github.com/Automattic/newspack-blocks/commit/598591168ada47998bada7ad97997e8ecbd5262c) to address this. Originally, the pattern was inserted immediately after the selected block. With this change, if the selected block is an empty paragraph, the pattern will replace it. I wasn't able to replicate the column block selection issue, but it's possible this will fix it too.
> 
> @thomasguillot other than this I think we're good to go. If you can verify this fix, I'll approve and we can merge.

@jeffersonrabb looks like you fixed it!!

cc @laurelfulford

**Pull request merged**